tardis/energy_input/util.py

```diff
@@ -0,0 +1,333 @@
+             GXPacket direction vector (unitary)
+    energy : float64
+             GXPacket energy
+    status : InteractionType
+             GXPacket status
+    shell : int64
+             GXPacket shell location index
+    """
+
+    def __init__(self, location, direction, energy, status, shell):
+        self.location = location
+        self.direction = direction
+        self.energy = energy
+        self.status = status
+        self.shell = shell
+        self.time_created = 0
+        self.time_current = 0
+        self.tau = -np.log(np.random.random())
+
+
+class SphericalVector(object):
+    """
+    Direction object to hold spherical polar and Cartesian directions
+    Must be initialized with r, theta, phi
+
+    Attributes
+    ----------
+    r : float64
+        vector r position
+    theta : float64
+        vector mu position
+    phi : float64
+        vector phi position
+    mu : float64
+        calculated vector theta position
+    x : float64
+        calculated vector x position
+    y : float64
+        calculated vector y position
+    z : float64
+        calculated vector z position
+    """
+
+    def __init__(self, r, theta, phi=0.0):
+        self.r = r
+        self.theta = theta
+        self.phi = phi
+
+    @property
+    def mu(self):
+        return np.cos(self.theta)
+
+    @property
+    def cartesian_coords(self):
+        # 0.5*np.pi subtracted because of the definition of theta
+        # in astropy.coordinates.cartesian_to_spherical
+        x, y, z = spherical_to_cartesian(
+            self.r, self.theta - 0.5 * np.pi, self.phi
+        )
+        return x.value, y.value, z.value
+
+
+def kappa_calculation(energy):
+    """
+    Calculates kappa for various other calculations
+    i.e. energy normalized to electron rest energy
+    511.0 KeV
+
+    Parameters
+    ----------
+    energy : float
+
+    Returns
+    -------
+    kappa : float
+
+    """
+    k = energy / 511.0
+    return k
+
+
+def euler_rodrigues(theta, direction):
+    """
+    Calculates the Euler-Rodrigues rotation matrix
+
+    Parameters
+    ----------
+    theta : float
+    direction : SphericalVector object
+
+    Returns
+    -------
+    rotation matrix : Two dimensional Numpy array, dtype float
+
+    """
+    a = np.cos(theta / 2)
+    dir_x, dir_y, dir_z = direction
+    b = dir_x * np.sin(theta / 2)
+    c = dir_y * np.sin(theta / 2)
+    d = dir_z * np.sin(theta / 2)
+
+    er11 = a ** 2.0 + b ** 2.0 - c ** 2.0 - d ** 2.0
+    er12 = 2.0 * (b * c - a * d)
+    er13 = 2.0 * (b * d + a * c)
+
+    er21 = 2.0 * (b * c + a * d)
+    er22 = a ** 2.0 + c ** 2.0 - b ** 2.0 - d ** 2.0
+    er23 = 2.0 * (c * d - a * b)
+
+    er31 = 2.0 * (b * d - a * c)
+    er32 = 2.0 * (c * d + a * b)
+    er33 = a ** 2.0 + d ** 2.0 - b ** 2.0 - c ** 2.0
+
+    return np.array(
+        [[er11, er12, er13], [er21, er22, er23], [er31, er32, er33]]
+    )
+
+
+def solve_quadratic_equation(x, y, z, x_dir, y_dir, z_dir, radius_velocity):
```

Comment:
maybe more similar to this https://github.com/tardis-sn/tardis/blob/master/tardis/montecarlo/montecarlo_numba/calculate_distances.py#L21 - also in name